IMP-/Small_StepT.thy

```diff
@@ -0,0 +1,179 @@
+where
+Assign:  "(x ::= a, s) \<rightarrow> (SKIP, s(x := aval a s))" |
+
+Seq1:    "(SKIP;;c\<^sub>2,s) \<rightarrow> (c\<^sub>2,s)" |
+Seq2:    "(c\<^sub>1,s) \<rightarrow> (c\<^sub>1',s') \<Longrightarrow> (c\<^sub>1;;c\<^sub>2,s) \<rightarrow> (c\<^sub>1';;c\<^sub>2,s')" |
+
+IfTrue:  "s b \<noteq> 0 \<Longrightarrow> (IF b \<noteq>0 THEN c\<^sub>1 ELSE c\<^sub>2,s) \<rightarrow> (c\<^sub>1,s)" |
+IfFalse: "s b = 0 \<Longrightarrow> (IF b \<noteq>0  THEN c\<^sub>1 ELSE c\<^sub>2,s) \<rightarrow> (c\<^sub>2,s)" |
+
+WhileTrue:   "s b \<noteq> 0 \<Longrightarrow> (WHILE b\<noteq>0 DO c,s) \<rightarrow>
+            (c ;; WHILE b \<noteq>0 DO c,s)" |
+WhileFalse:   "s b = 0 \<Longrightarrow> (WHILE b\<noteq>0 DO c,s) \<rightarrow>
+            (SKIP,s)"
+
+subsection "Transitive Closure"
+abbreviation
+  small_step_pow :: "com * state \<Rightarrow> nat \<Rightarrow> com * state \<Rightarrow> bool" ("_ \<rightarrow>* _ \<down> _" 55)
+  where "x \<rightarrow>* t \<down> y == (rel_pow  small_step t)  x y"
+
+
+subsection\<open>Executability\<close>
+
+code_pred small_step .
+
+values "{(c',map t [''x'',''y'',''z'']) |c' t .
+   ((''x'' ::= A (V ''z'');; ''y'' ::=A ( V ''x''),
+    <''x'' := 3, ''y'' := 7, ''z'' := 5>) \<rightarrow>* 1 \<down> (c',t))}"
+subsection\<open>Proof infrastructure\<close>
+
+subsubsection\<open>Induction rules\<close>
+
+text\<open>The default induction rule @{thm[source] small_step.induct} only works
+for lemmas of the form \<open>a \<rightarrow> b \<Longrightarrow> \<dots>\<close> where \<open>a\<close> and \<open>b\<close> are
+not already pairs \<open>(DUMMY,DUMMY)\<close>. We can generate a suitable variant
+of @{thm[source] small_step.induct} for pairs by ``splitting'' the arguments
+\<open>\<rightarrow>\<close> into pairs:\<close>
+lemmas small_step_induct = small_step.induct[split_format(complete)]
+
+
+subsubsection\<open>Proof automation\<close>
+
+declare small_step.intros[simp,intro]
+
+
+text\<open>Rule inversion:\<close>
+
+inductive_cases SkipE[elim!]: "(SKIP,s) \<rightarrow> ct"
+thm SkipE
+inductive_cases AssignE[elim!]: "(x::=a,s) \<rightarrow> ct"
+thm AssignE
+inductive_cases SeqE[elim]: "(c1;;c2,s) \<rightarrow> ct"
+thm SeqE
+inductive_cases IfE[elim!]: "(IF b\<noteq>0 THEN c1 ELSE c2,s) \<rightarrow> ct"
+inductive_cases WhileE[elim]: "(WHILE b\<noteq>0 DO c, s) \<rightarrow> ct"
+thm WhileE
+text\<open>A simple property:\<close>
+lemma deterministic:
+  "cs \<rightarrow> cs' \<Longrightarrow> cs \<rightarrow> cs'' \<Longrightarrow> cs'' = cs'"
+apply(induction arbitrary: cs'' rule: small_step.induct)
+        apply blast+
+  apply auto
+done
+
+section "Equivalence with big-step semantics"
+declare rel_pow_intros[simp,intro]
+text "sequence property"
+lemma star_seq2: "(c1,s) \<rightarrow>* t \<down> (c1',s') \<Longrightarrow> (c1;;c2,s) \<rightarrow>* t  \<down> (c1';;c2,s')"
+proof(induction t arbitrary: c1 c1' s s')
+  case 0
+  then show ?case by auto
+next
```

Comment:
for easy cases that are solved by auto, you can leave the case out and close the proof with `qed auto`, which will use the tactic on the remaining cases

Reply:
done